### PR TITLE
fix: Garbage collector hangs indefinitely on shutdown

### LINKED ISSUE #7558: garbage collector service hangs indefinitely on shutdown
__Steps to reproduce:__

Run this command
```shell
LOG_FILTER=trace RUST_LOG=trace influxdb_iox  run  garbage-collector --catalog=memory
```

And then send SIGINT (aka hit CTRL-C) 

__Expected behaviour:__
I expect the garbage collector to shutdown cleanly

__Actual behaviour:__
The garbage collector hangs indefinitely (and I can only stop it via SIGKILL)
```shell
$ killall -KILL influxdb_iox
```

__Environment info:__
mac x86

__Logs:__

```
(arrow_dev) alamb@MacBook-Pro-8:~/Software/influxdb_iox2$ LOG_FILTER=trace RUST_LOG=trace influxdb_iox  run  garbage-collector --catalog=memory
2023-04-14T14:21:51.120376Z  INFO clap_blocks::object_store: Object Store object_store_type="Memory"
2023-04-14T14:21:51.121538Z  INFO influxdb_iox::commands::run::garbage_collector: starting garbage-collector
2023-04-14T14:21:51.123379Z  INFO influxdb_iox::commands::run::main: InfluxDB IOx server starting git_hash=46ca92ece4a0ccbb62f2a6e95681346020f4cab8 version=0.1.0 uuid=32bf1c2e-d617-4475-8346-b7a811489717 num_cpus=16 build_malloc_conf=
2023-04-14T14:21:51.125116Z  INFO garbage_collector: GarbageCollector starting objectstore_cutoff_days=14days parquetfile_cutoff_days=14days objectstore_sleep_interval_minutes=30 parquetfile_sleep_interval_minutes=30 retention_sleep_interval_minutes=35
2023-04-14T14:21:51.125655Z DEBUG influxdb_iox::commands::run::main: wait for handle server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:51.126411Z  INFO influxdb_iox::commands::run::main: Binding gRPC services grpc_bind_address=SocketAddr(127.0.0.1:8082) server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:51.126758Z TRACE mio::poll: registering event source with poller: token=Token(0), interests=READABLE | WRITABLE
2023-04-14T14:21:51.127068Z  INFO ioxd_common: bound gRPC listener local_addr=127.0.0.1:8082
2023-04-14T14:21:51.127420Z  INFO influxdb_iox::commands::run::main: Completed bind of gRPC, binding http http_bind_address=SocketAddr(127.0.0.1:8080) server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:51.127733Z TRACE mio::poll: registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
2023-04-14T14:21:51.127747Z  INFO ioxd_common: bound HTTP listener bind_addr=127.0.0.1:8080
2023-04-14T14:21:51.127759Z  INFO ioxd_common: gRPC server listening server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:51.127763Z  INFO ioxd_common: InfluxDB IOx Server(GarbageCollector) { .. } server ready
2023-04-14T14:21:51.136400Z  INFO ioxd_common: HTTP server listening server_type=Server(GarbageCollector) { .. }
^C2023-04-14T14:21:55.412355Z  INFO ioxd_common: Received SIGINT
2023-04-14T14:21:55.412392Z  INFO ioxd_common: shutdown requested server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:55.412412Z DEBUG hyper::server::shutdown: signal received, starting graceful shutdown
2023-04-14T14:21:55.412433Z TRACE mio::poll: deregistering event source from poller
2023-04-14T14:21:55.413691Z  INFO ioxd_common: gRPC server shutdown server_type=Server(GarbageCollector) { .. }
2023-04-14T14:21:55.413710Z  INFO ioxd_common: frontend shutdown completed server_type=Server(GarbageCollector) { .. }
Killed: 9
```

### PR DESCRIPTION
**Pull request by aprimadi**
Closes #7558 

The problem is that the objectstore lister task sleeps without listening to shutdown signal.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Review by domodwyer — COMMENTED** [review]
Hey @aprimadi - thanks for the PR! :heart:

This looks great, I'd love to merge it - there's a few CI steps failing though ([lint failures] and needs a `cargo fmt`)

[lint failures]:https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/32298/workflows/d17a9469-57a9-4110-92b2-51e716f4fee5/jobs/298339

<details><summary>Lint failures</summary>
<p>

```
error: using `.clone()` on a ref-counted pointer
  --> garbage_collector/src/objectstore/deleter.rs:79:44
   |
79 |         let items = populate_os_with_items(object_store.clone(), nitems).await;
   |                                            ^^^^^^^^^^^^^^^^^^^^ help: try this: `Arc::<object_store::memory::InMemory>::clone(&object_store)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr
   = note: `-D clippy::clone-on-ref-ptr` implied by `-D warnings`

error: using `.clone()` on a ref-counted pointer
  --> garbage_collector/src/objectstore/deleter.rs:81:37
   |
81 |         assert_eq!(count_os_element(object_store.clone()).await, nitems);
   |                                     ^^^^^^^^^^^^^^^^^^^^ help: try this: `Arc::<object_store::memory::InMemory>::clone(&object_store)`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr

error: using `.clone()` on a ref-counted pointer
   --> garbage_collector/src/objectstore/deleter.rs:112:13
    |
112 |             object_store.clone(),
    |             ^^^^^^^^^^^^^^^^^^^^ help: try this: `Arc::<object_store::memory::InMemory>::clone(&object_store)`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_ref_ptr

error: this let-binding has unit value
   --> garbage_collector/src/objectstore/deleter.rs:117:9
    |
117 | /         let res = tokio::time::timeout(Duration::from_secs(3), perform_fu)
118 | |             .await.unwrap().unwrap();
    | |_____________________________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
    = note: `-D clippy::let-unit-value` implied by `-D warnings`
help: omit the `let` binding
    |
117 ~         tokio::time::timeout(Duration::from_secs(3), perform_fu)
118 +             .await.unwrap().unwrap();
    |

error: `assert_eq` of unit values detected. This will always succeed
   --> garbage_collector/src/objectstore/deleter.rs:119:9
    |
119 |         assert_eq!(res, ());
    |         ^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unit_cmp
    = note: `#[deny(clippy::unit_cmp)]` on by default

error: returning the result of a `let` binding from a block
   --> garbage_collector/src/objectstore/deleter.rs:125:9
    |
124 |         let count = objects.fold(0, |acc, _| async move { acc + 1 }).await;
    |         ------------------------------------------------------------------- unnecessary `let` binding
125 |         count
    |         ^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
    = note: `-D clippy::let-and-return` implied by `-D warnings`
help: return the expression directly
    |
124 ~         
125 ~         objects.fold(0, |acc, _| async move { acc + 1 }).await
    |
```

</p>
</details>

**Comment by aprimadi** [comment]
@domodwyer fixed. Sorry I was waiting for some feedback.

**Review by domodwyer — APPROVED** [review]
@aprimadi - no feedback needed - it's perfect! 👌 

Thank you for the PR!

File: garbage_collector/src/objectstore/deleter.rs [hunk 1]
```diff
@@ -47,3 +58,98 @@ pub enum Error {
 }
 
 pub(crate) type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[cfg(test)]
```
Comment (alamb) [inline garbage_collector/src/objectstore/deleter.rs:62]: ❤️  what I lovely test . Thank you

**Pull request merged**